specs/request-handler-plugins.md

```diff
@@ -0,0 +1,116 @@
+# Support a "plug-in" model for request handlers
+
+To allow developers to provide unique handling behaviors for different APIs the Microsoft Graph will have a plug-in model to dynamically load request handlers. This will provide separation of concerns between the software components intercepting network request and those providing responses.
+To enable developers building handlers an abstractions library will be distributed on NuGet.
+
+## History
+
+| Version | Date | Comments | Author |
+| ------- | -------- | ----- | --- |
+| 1.0 | 2022-12-19 | Initial specifications | @gavinbarron |
+
+## Configuration file
+
+The existing `appsettings.json` file will be used to provide configuration. Both for determining which request handlers are loaded into the proxy and providing configuration for each of the loaded request handlers. The set of request handlers to be loaded will be specified as an array of objects at the root of the configuration file in the `handlers` property. Each handler must supply a `configSection`, `handerPath`, `name`, and `urlsToWatch` in the handlers array.
```

Comment:
What if we automatically loaded plugins from all dlls in the proxy folder? Folks would only need to specify them in the config file, if the plugin requires config. When they don't need a config, they can just drop the dll and off they go.

Reply:
I looked at that idea, it's certainly an option, the biggest risk/drawback is that if there are two plugins with conflicting dependencies it won't work. Imagine one plugin that uses Newtonsoft.Json @ v8 and a second that uses Newtonsoft.Json @ v13. Loading the assemblies directly out of the same folder would cause the app to crash when it failed to resolve one of the version of Newtonsoft.Json

Reply:
Good point. Also, we'd have no way to determine their order. Let's stick with your proposal.